arangod/Pregel/Algos/AIR/Greenspun/Interpreter.cpp

```diff
@@ -0,0 +1,773 @@
+    return res;
+  }
+  if (!proto.slice().isNumber()) {
+    return EvalError("expected numeric expression in pattern");
+  }
+  auto pattern = proto.slice().getNumber<double>();
+  paramIterator++;
+  for (; paramIterator.valid(); paramIterator++) {
+    auto pair = *paramIterator;
+    if (!pair.isArray() || pair.length() != 2) {
+      return EvalError("in case " + std::to_string(paramIterator.index()) +
+                       ", expected pair, found: " + pair.toJson());
+    }
+    auto&& [cmp, body] = unpackTuple<VPackSlice, VPackSlice>(pair);
+    VPackBuilder cmpValue;
+    if (auto res = Evaluate(ctx, cmp, cmpValue); !res) {
+      return res.mapError([&](EvalError& err) {
+        err.wrapMessage("in condition " + std::to_string(paramIterator.index() - 1));
+      });
+    }
+
+    if (!cmpValue.slice().isNumber()) {
+      return EvalError("in condition " + std::to_string(paramIterator.index() - 1) +
+                       " expected numeric value, found: " + cmpValue.toJson());
+    }
+
+    if (pattern == cmpValue.slice().getNumber<double>()) {
+      return Evaluate(ctx, body, result).mapError([&](EvalError& err) {
+        err.wrapMessage("in case " + std::to_string(paramIterator.index() - 1));
+      });
+    }
+  }
+
+  result.add(VPackSlice::noneSlice());
+  return {};
+}
+
+EvalResult SpecialForEach(Machine& ctx, ArrayIterator paramIterator, VPackBuilder& result) {
+  // ["for-each", ["a", ["A", "B", "C"]], ["d", ["1", "2", "3"]], ["print", ["var-ref", "a"], ["var-ref", "d"]]]
+
+  if (!paramIterator.valid()) {
+    return EvalError("Expected at least one argument");
+  }
+
+  struct IteratorTriple {
+    std::string_view varName;
+    VPackBuilder value;
+    ArrayIterator iterator;
+
+    IteratorTriple(std::string_view name, VPackBuilder v)
+        : varName(name), value(std::move(v)), iterator(value.slice()) {}
+  };
+
+  std::vector<IteratorTriple> iterators;
+
+  auto const readIteratorPair = [&](VPackSlice pair) -> EvalResult {
+    if (pair.length() != 2) {
+      return EvalError("Expected pair, found: " + pair.toJson());
+    }
+    auto&& [var, array] = unpackTuple<VPackSlice, VPackSlice>(pair);
+    if (!var.isString()) {
+      return EvalError("Expected string as first entry, found: " + var.toJson());
+    }
+    if (!array.isArray()) {
+      return EvalError("Expected array os second entry, found: " + array.toJson());
+    }
+    VPackBuilder listResult;
+    if (auto res = Evaluate(ctx, array, listResult); res.fail()) {
+      return res;
+    }
+
+    iterators.emplace_back(var.stringView(), std::move(listResult));
+    return {};
+  };
+
+  while (!paramIterator.isLast()) {
+    VPackSlice pair = *paramIterator;
+    paramIterator++;
+    if (auto res = readIteratorPair(pair); res.fail()) {
+      return res.mapError([&](EvalError& err) {
+        err.wrapMessage("at position " + std::to_string(paramIterator.index() - 1));
+      });
+    }
+  }
```

Comment:
You could use a `for (; !paramIterator.isLast(); ++paramIterator)` loop, and get rid of the offset (`index() - 1`) on the way.